teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,330 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+## Proposal
+
+This TEP provides a plan for ensuring that feature stability doesn't depend on CRD API version, and using per-feature flags to move to a future where only stable features are enabled by default through the following steps:
+- [Change existing validation to decouple feature and API versioning](#change-existing-validation-to-decouple-feature-and-api-versioning)
+- [Per feature flag for new api-driven features](#per-feature-flag-for-new-api-driven-features)
+- [Sunset `enable-api-fields` after existing features stabilize](#sunset-enable-api-fields-after-existing-features-stabilize)
+
+
+## Design Details
+
+### Change existing validation to decouple feature and API versioning
+We will change the current validation for `enable-api-fields=stable` to only allow using stable features regardless of API version. This will resolve the current issue of the coupling of API and feature versioning in v1beta1. More specifically, beta features resolvers, object/ array params and results will require `enable-api-fields` set to `beta` to be used. This means that, for those current beta features (resolvers, object/ array params and results) will no longer be enabled with `enable-api-fields=stable`. 
+To continue using these features, users will need to explicitly set the `enable-api-fields` flag to `beta`. This change will not affect users who are already using the `enable-api-fields=beta` flag, which is the default and will continue to be. It would change the behaviour for those who only want to enable `stable` features. 
+
+Note that although this is a behavior change, it is more of a bug fix for the coupling of feature and API versioning. Currently, with `enable-api-fields` set to `stable`,  PipelineRuns like [this one](https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/beta/git-resolver.yaml) fail because the controller cannot create child TaskRuns. This change will result in a validation failure instead.
+
+However, the default value of `enable-api-fields` will **not** be changed and it will remain to be `beta`, so the **beta** features will continue to be enabled by default.
+
+- **Impacts on users:**
+    - Cluster operators:
+      - Current cluster operators with `enable-api-fields` set to `alpha` or `beta` should not experience any changes.
+      - This makes it possible for cluster operators to have full control over only stable feature usages, rather than user overrides. For example, currently if cluster operators want their users to only opt-in “stable” features, they cannot do so for v1beta1 apiVersion for the exception of resolver, object params and results.
+    - Pipeline and Task authors: 
+    This affects pipeline users in the situation where cluster operators have chosen to set `enable-api-fields` to `stable` (i.e. those who have changed the current default `beta` value) and who are accidentally using beta features, like resolvers, in v1beta1.
```

Comment:
I am generally not in favor of using this term (`accidentally`) here, specially in the TEP proposal.

Let's try to explain what this accident mean here.

`v1beat1` Pipeline definitions using `resolver` is applied to a cluster with `enable-api-fields` set to `stable`, is it? But this `pipeline` does not result in a successful `pipelineRun` because of <some reason>.  The resolver is still a `beta` feature and this pipeline should not have been applied to the cluster in first place. Does this make sense?

Reply:
Thanks Priti! I've added the examples and reasoning for elaborating this.